### PR TITLE
[fix][client] Fix producer could send timeout when enable batching

### PR DESCRIPTION
**Pull request by AnonHxy**
Fixes https://github.com/apache/pulsar/issues/19189

### Motivation

Fixes https://github.com/apache/pulsar/issues/19189

The root cause is that:
1. When msg size is greater than `batchingMaxBytes `,  we will go to line677:
https://github.com/apache/pulsar/blob/4b0dc9ab87b5fa82525514dc37a91e7db0fa95fc/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L645-L678

2. Let go into the method `doBatchSendAndAdd( )`：
    2.1 Line882 do nothing and will not trigger the batch flush task;
    2.2 Line883 just add the msg into `batchMessageContainer`;
    2.3 Note that, the batch flush task doesn't run now. And if there is no successive messages, the msg in `batchMessageContainer` will never be flushed to network

https://github.com/apache/pulsar/blob/4b0dc9ab87b5fa82525514dc37a91e7db0fa95fc/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L876-L888

To fix this issue, after adding the msg to the `batchMessageContainer` we need:
1.  trigger the batch flush task
2.  or send the msg dirrectly if the `batchMessageContainer` is full.
 

### Modifications

Trigger send if `batchMessageContainer` is full or schedule the batch flush task

### Verifying this change

- [x] Make sure that the change passes the CI checks.
- Add UT `org.apache.pulsar.client.api.SimpleProducerConsumerTest#testSendMsgGreaterThanBatchingMaxBytes`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/AnonHxy/pulsar/pull/26

### CODE REVIEW

**Review by Technoboy- — APPROVED** [review]

**Comment by HQebupt** [comment]
Nice catch!

**Review by HQebupt — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
Great catch!

**Comment by lhotari** [comment]
This is a good catch, but I didn't have a chance to review it before merging. 

@Technoboy- Please wait for reviews from the people who have worked on the code when such issues are faced.

**Comment by lhotari** [comment]
I don't fully agree on the fix. When looking at the logic, it feels that org.apache.pulsar.client.impl.AbstractBatchMessageContainer#haveEnoughSpace is wrong if a bug like this occurs. @codelipenghui @michaeljmarshall do you have a chance to check this PR although it has already been merged?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java
Comment (michaeljmarshall) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:889]: Should this be before the call to `triggerSendIfFullOrScheduleFlush`? It is in the other usage of `triggerSendIfFullOrScheduleFlush`. However, we're in a `synchronized (this)` block, so it might be fine.

**Review by michaeljmarshall — COMMENTED** [review]
I think this is the right change. Good catch @AnonHxy. Here is my analysis, so we can verify the change:

This change solves 2 problems.

1. The first problem is described by the issue/PR and the problem shown in the added test. This is the case where the message is large enough to trigger a batch send now.
2. The second problem is where `doBatchSendAndAdd` is called because of logic related to the `sequenceId`.

In both cases, we add a message to the `batchMessageContainer` without doing something to ensure that batch will get flushed. In the first case, we should flush the batch immediately and in the second case, we should start a timer to potentially batch more messages then flush the batch.

An alternate solution to the first problem is shown in the Go client: https://github.com/apache/pulsar-client-go/pull/792. There, we always let a single message get added to the batch. If we go in this direction, we'd still need to cover the second case.

It would have been nice to get a test to cover the second case.

**Pull request merged**
